### ISSUE TITLE
`data` constructors should *not* be non-linear unless defined within the scope of `with-K`

### ISSUE THREAD

**Issue by jamesmckinna** [agda/agda-stdlib#2519]
See the discussion on #2504 for a specific instance. 

This raises the issue to library level, given the ambient `stdlib` assumption of `cubical-compatible`/`without-K`.

`data` constructors which exploit non-linearity are making *implicit* appeal to reflexivity of (propositional) equality, and that should instead be required as an *explicit* witness. Specialising the constructor to supply `refl` as a default witness should be the preferred approach to offering a non-linear constructor. Cf. `Data.Nat.Divisibility.Core` #2013

**Comment by MatthewDaggitt** [comment]
I agree this is a good principle going forward. However, how many definitions fall foul of this currently? Depending on the numbers I would be weary of how big a breaking change this would be to change all the existing definitions?

**Comment by jamesmckinna** [comment]
I haven't done a survey of the relevant definitions, besides `Data.Nat.Base._≤′_`, which seems to fall even more easily to the corresponding refactoring as #2504 .

**Comment by JacquesCarette** [comment]
I agree with the design principle.

My gut feel is that this isn't likely to occur very often. But this is rather hard to search for! So piecemeal refactoring is indeed a rather reasonable way to go.

**Comment by Taneb** [comment]
Can you clarify what exactly you mean by "non-linear" here? Is it that the same variable should not appear in two different indices of a data constructor?

**Comment by jamesmckinna** [comment]
> Can you clarify what exactly you mean by "non-linear" here? Is it that the same variable should not appear in two different indices of a data constructor?

Yes, exactly. I'll change the title of the issue.

In the originating PR #2504 it's actually even stronger than that, because it meant that those indices could even become *parameters* to the `data` definition.

**Comment by jamesmckinna** [comment]
Looping back to #1344 ... where @MatthewDaggitt had the same idea, but with different motivations, in [this comment](https://github.com/agda/agda-stdlib/issues/1344#issuecomment-768686234).
